Ladder step2

HaeSeongPark opened the pull request:
step1에 대한 피드백 반영 및 step2 진행했습니다~

HaeSeongPark committed:
피드백과 2단계 요구사항 정리

HaeSeongPark committed:
강제 언랩핑 피드백 반영

HaeSeongPark committed:
printLadder함수 depth 줄이기

HaeSeongPark committed:
refactoring Scanner.userInputValue with step2 example

HaeSeongPark committed:
make LadderGame.swift file

HaeSeongPark committed:
make Input.swift file

HaeSeongPark committed:
change name to Input

HaeSeongPark committed:
make Output.swift file

HaeSeongPark committed:
input 객체 완료

HaeSeongPark committed:
ladder 객체 완료

HaeSeongPark committed:
output 객체 완료

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -27,11 +25,11 @@ struct LadderGame {
     var heightOfLadder:Int = 0
-    var ladderMatrix = [[Bool]]()
+    var ladderMatrix = [[String]]()
     let printer:Printer = Printer()
```
ladderMatrix를 Bool 에서 String 으로 바꾼 이유가 궁금합니다.

HaeSeongPark replied:
Bool로 하면
if else로 true false인지 체크하는 과정을 추가해야 하는데 그러면 depth가 늘어서나서
depth를 줄이고자 String으로 했습니다

godrm replied:
코드는 줄어들지만 사다리를 표시하는 방법이 '-' 에서 '---'로 바뀌면 어떻게 될까요? 
- 저장하는 데이터 구조와 화면에 출력하는 데이터 구조가 항상 일치해야할까요? 
- String 과 Bool 타입의 메모리 레이아웃 크기 차이는 얼마나 날까요?

godrm commented on LadderGame/LadderGame/Input.swift:
```diff
@@ -0,0 +1,50 @@
+    private var numberOfPerson:Int = 0
+    private var heightOfLadder:Int = 0
+    private func inValidCheck(_ inputValue:String?) -> Int {
+        guard inputValue != nil || (inputValue?.isEmpty ?? false) else {
```
guard let 으로 개선할 수 있을까요?

HaeSeongPark replied:
guard let 으로 개선 했습니다.

godrm commented on LadderGame/LadderGame/Input.swift:
```diff
@@ -0,0 +1,50 @@
+        if rawHeightOfPerson != 0 {
+            heightOfLadder = rawHeightOfPerson
+        }
+    }
```
start() 함수에서는 어떤 동작을 하고 있나요?
- 참여자 수를 입력받는 것 동작과 사다리 높이를 입력받는 동작 두 가지를 하고 있네요.
- 함수 하나가 한 가지 동작만 하도록 개선해보세요.
- 참여자나 사다리 높이를 0을 입력하면 어떻게 되는건가요?

HaeSeongPark replied:
start()를 함수를 askNumberOfPerson() 과 askHeightOfLadder()으로 나눴습니다
참여자나 사다리 높이를 0을 입력하면 어떻게 되는건가요? -> 0이면 다시 입력하게 했습니다

godrm commented on LadderGame/LadderGame/Input.swift:
```diff
@@ -0,0 +1,50 @@
+        }
+    }
+    var result: (Int,Int){
+        return (numberOfPerson, heightOfLadder)
```
이런 변수를 사용하기 보다 start() 함수에서 return 값으로 사용하면 어떨까요?

HaeSeongPark replied:
start()가 삭제 후 askNumberOfPerson과 askHeightOfLadder에서 리턴하게끔 했습니다.
그 과정에서 result와 numberOfPerson변수와 heightOfLadder변수가 필요없어져서 삭제했습니다

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -1,119 +1,26 @@
+    var input = Input()
+    input.start()
+    let inputValue:(numerOfPerson:Int, heightOffLadder:Int) = input.result
+    var ladderGame = LadderGame(inputValue.numerOfPerson, inputValue.heightOffLadder)
```
`inputValue:(numerOfPerson:Int, heightOffLadder:Int) ` 튜플을 썼으면
LadderGame(inputValue)처럼 간단하게 넘기는것도 가능하지 않나요? 어떤 차이가 있나요?

HaeSeongPark replied:
아 그렇군요... 위에서 수정하면서 사람과 사다리높이를 각각 가져와서 각각 넣었습니다!

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -1,119 +1,26 @@
+func main(){
+    var input = Input()
+    input.start()
+    let inputValue:(numerOfPerson:Int, heightOffLadder:Int) = input.result
```
inputValue 라는 변수명 보다는 좀더 구체적인 역할을 표현하는 단어로 바꿔보세요.

HaeSeongPark replied:
수정하면서 numberOfPerson와 heightOfLadder로 바뀌었습니다.

godrm commented on LadderGame/LadderGame/LadderGame.swift:
```diff
@@ -0,0 +1,47 @@
+        for _ in 0..<numberOfLadder{
+            ladderMatrix[heightOfLadder.arc4random][(numberOfPerson-1).arc4random] = "-|"
+        }
+        return ladderMatrix
```
여기서 배열을 바로 리턴할꺼라면 ladderMatrix 인스턴스 변수가 꼭 필요한가요? 
로컬 변수만 있어서 되지 않나요?

HaeSeongPark replied:
아하... 근데 인스턴스 변수보다 로컬변수를 쓰는 게 왜 좋은건가요?

godrm replied:
인스턴스 변수는 인스턴스가 메모리에 있는 동안 메모리에 남아있지만,
로컬 변수는 해당 메소드만 동작한 이후에 사라지겠죠. 
메소드 입장에서도 인스턴스 변수의 값에 따라 동작이 달라지기 보다는
입력값에 따라서 리턴값이 결정되는 순수한 함수로 만드는 게 좋습니다.

godrm commented on LadderGame/LadderGame/LadderGame.swift:
```diff
@@ -0,0 +1,47 @@
+    mutating func makeLadder() -> [[String]]{
+        let numberOfLadder = ( numberOfPerson + heightOfLadder ) / 2
+        for _ in 0..<numberOfLadder{
+            ladderMatrix[heightOfLadder.arc4random][(numberOfPerson-1).arc4random] = "-|"
```
이렇게 사다리를 만들면 `|---|---|` 이렇게 나란히 생기는 사다리가 나오지 않을까요?

HaeSeongPark replied:
네 랜덤으로 생성 된 숫자의 해당 위치에 단지 추가하는 것이기 때문에 나란히는 생기지 않을 것 같습니다.

godrm replied:
말한것처럼 당장은 한 위치에만 생기지만, 반복문으로 돌면서 만들다보면 생기지 않냐는 질문입니다.

godrm reviewed (commented):
객체 역할 분담은 좋아졌네요. 피드백을 확인해보세요.

HaeSeongPark committed:
inValidCheck guard let으로 개선, start함수 askNumberOfPerson과 askHeightOfLadder()로 나누고 0이면 재귀처리

HaeSeongPark committed:
askNumberOfPerson에서 numbrOfPerson을 askHeightOfLadder에서 heightOfLadder를 리턴
그 과정에서 result와 numberOfPerson변수와 heightOfLadder변수가 필요없어져서 삭제

HaeSeongPark committed:
ladderMatrix를 makeLadder함수의 지역변수로 수정

godrm commented on LadderGame/LadderGame/Input.swift:
```diff
@@ -0,0 +1,56 @@
+        if numberOfPerson != 0 {
+            return numberOfPerson
+        }
+        return askNumberOfPesron()
```
0을 입력하면 다시 입력받는 것을 재귀로 호출하는군요!
어떤 함수를 호출하고 반복해서 호출할 지 판단하는 흐름 제어는 그 함수를 호출한 상위 모듈에서 하는게 좋습니다.
하위 모듈에서 무한 반복이 되도록 재귀 호출을 하는 방법은 지양하세요.

godrm commented on LadderGame/LadderGame/Output.swift:
```diff
@@ -0,0 +1,27 @@
+            print(row.joined(separator: ""), terminator:"")
+            print()
+        }
+    }
```
이렇게 인스턴스 변수가 없이 순수한 함수만 있는 경우는
static 을 붙여서 타입 메소드로 만들어도 좋습니다.

godrm reviewed (commented):
수고하셨습니다. 이 상태로 다음 단계도 진행 가능할 것 같습니다.
재귀 호출 관련된 피드백을 다음 단계 미션을 하면서 함께 개선해보세요.
